### ISSUE TITLE
Support for testing non-package R project structures

### ISSUE THREAD

**Issue by telegott** [r-lib/testthat#1490]
This was already [posted](https://github.com/r-lib/testthat/issues/916) a while ago, but I'm not sure that the poster's point was understood.
In analogy to how `lintr` or `styler` can recursively process directories, it would be great to see `testthat` supporting something like a `recursive=TRUE` option for `testthat::test_dir` that reports failures along the way and has some bool that can be used for determining the exit code of a shell script for use in CI.
I tried to hack something together myself by looking at the `testthat` code, but it didn't end up being satisfactory, also stackoverflow answers were pretty "hand-made". 

I would find such a feature highly useful. I know that `testthat` is geared for testing packages, but the reality I often see is that despite the many helper packages, it's a large burden for even longer time R users. 

Packages like `renv` which in my understanding are a bit redundant in the case of package development since `DESCRIPTION` more or less does the same seem to acknowledge the use case of R outside of a package structure. Things like the `box` package and Shiny modules seem to steer in the same direction, and they allow for more modularity in terms of naming and arranging code (subdirectories in `R/`), which cannot be tested with `testthat` currently in a non-hackish manner if I'm informed correctly.

In summary, I think this feature would make testing much more accessible for many users for which currently there's the decision is between developing a package and testing or stick to known structures, in which case starting with good practices (e.g. testing) is not easy.

**Comment by hadley** [comment]
Unfortunately I don't find this argument compelling because the necessary "package" structure to use testthat is so minimal — `DESCRIPTION` at the top-level (to say what packages are needed to run the test), R code in `R`, and tests in `tests/testthat`. All of these pieces are required for testing, so in my mind what you're asking for is some way to specify that those pieces live in different places. You'd presumably want to store that information in a file (so you didn't have to re-enter it every time you ran the tests) so now we have to define where that file lives and what format it uses etc etc.
